### ISSUE TITLE
TreeGrid with expanded nodes flickers on each refresh of the data

### ISSUE THREAD

**Issue by wofl** [vaadin/flow-components#1981]
### Description

The TreeGrid recursively rebuilds the tree on any refresh of the underlying data. This leads to really annoying flickering because each intermediate state is rendered to the client and even more important to a incredible performance penalty.
The dead simple example on the official vaadin [homepage ](https://vaadin.com/components/vaadin-tree-grid/) takes almost 2 seconds when all nodes are expanded (just 15 rows!) and somebody clicks on any header to sort it.

### Expected outcome

The TreeGrid makes one request and shows the result or atleast only renders the final state to the client without those intermediate steps.

### Actual outcome

All intermediate states are rendered, the TreeGrid flickers and takes a disproportionate amount of time to refresh.

### Live Demo

The first TreeGrid here:
https://vaadin.com/components/vaadin-tree-grid/java-examples
with all nodes expanded and a click on any header.

### Browsers Affected
- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge
- [ ] IE 11
- [ ] iOS Safari
- [ ] Android Chrome

**Comment by TatuLund** [comment]
There is actually better example here: https://cookbook.vaadin.com/treegrid-subtree-highlight

Open nodes and select a row in the demo, that will call refreshAll, and you will see the flicker.

**Comment by wofl** [comment]
If you see #1050 as a duplicate, IMHO the title and description there is misleading. The expanded/collapsed state **is** retained and because of this and the recursive behaviour of the TreeGrid the flicker occurs.

**Comment by TatuLund** [comment]
>  The expanded/collapsed state is retained and

Good point, what I am meaning in the other ticket is that it is not retained during the process, thus the flicker.

**Comment by wofl** [comment]
Additionaly there is no word about the high performance penalty, is there? I mean - making 6 requests and taking 2 seconds for just showing the same data in a different order seems really odd.

**Comment by TatuLund** [comment]
@wofl The performance penalty is also correct, but that is even wider topic, which is dealt in ticket https://github.com/vaadin/flow-components/issues/1627

**Comment by TatuLund** [comment]
The problem will be fixed in Vaadin 23.1. Meanwhile as a workaround you can extend TreeGrid and use the code in this patch proposal https://github.com/vaadin/flow-components/pull/2712/files

**Comment by stefanuebe** [comment]
Also in Vaadin 23.3 this is still an issue, when having huge trees and scroll somewhere down the tree (e.g. to the mid). Still flickers a lot around. Indeed, it works nice with small trees and/or when beingt at the top.

**Comment by vursen** [comment]
Closing as resolved by https://github.com/vaadin/platform/issues/7843 in Vaadin 25. The client now receives the visible range as a flat list in a single request. Feel free to reopen the ticket with a reproduction example if the issue still persists.
